README.md
<img src="input/autoMP.png" width="500"/>\
A tool for converting images into a Lua script for rendering in Mario Paint.

autoMP by marl

Based on Automatic Mario Paint by alden

Logo artwork, readme and extra help by G-Zone

Special thanks to greysondn

*This guide is written with accessibility in mind. My hope is that anyone will be able to use autoMP, even if you, like myself, are unfamiliar with things like Python and related software.  ~G-Zone*

**These instructions assume you are using Windows 10.**


## What you need

#### autoMP
Our image converting and script writing utility. Be sure to download the latest version!\
-> https://github.com/marl64/autoMP/releases

#### Python
The interpreter for the Python programming language, which autoMP is built with.
* In the installer, be sure that "Add Python to PATH" is checked on the first page, then click Customize installation.
* For Optional Features, leave everything checked.
* For Advanced Options, be sure to check "Decompile standard library".\
-> https://www.python.org/downloads/

#### Pillow
A set of libraries for Python, required to run autoMP.
* After installing Python as instructed above, open Command Prompt (Press the Windows key, then type cmd), then right-click and press "Run as administrator".
* Type "pip install --upgrade Pillow" and press enter.

#### SNES9x with Lua support
This is a version of the SNES9x emulator which lets you run game-altering scripts.\
-> https://github.com/TASVideos/snes9x-rr/releases/tag/snes9x-151-v7.1

#### Lunar IPS
This is a very simple utility for patching ROM files.\
-> https://fusoya.eludevisibility.org/lips/

#### Mario Paint ROM file
Mario Paint is a 1992 art program for the Super Nintendo Entertainment System/Super Famicom that came bundled with the SNES Mouse.\
-> We can't provide a source for this file. You'll have to either dump an official cartridge yourself, or obtain it some other way. A quick web search should prove fruitful.


## Setup

1. Download and extract the autoMP source code.

2. If they don't already exist, create 2 new folders, "input" and "output" in the autoMP folder.

3. Place the image(s) you want to use into the input folder. Pretty much any image file should work. If one isn't accepted, try saving it as a common filetype, like .jpg or .png.

4. Double-click autoMP.py. This should fill the output folder with preview images, as well as lua scripts for each image in your input folder.

5. Open Lunar IPS. Click "Apply IPS Patch" and browse for the patch we provided, MarioPaintJUh1Joystick.ips, then for the Mario Paint ROM you want to patch it to. This modifies your ROM to accept the gamepad rather than the mouse for inputs, which is imperative for autoMP to function properly.

6. Extract the SNES9x emulator into a seperate folder from autoMP. Place the provided snes9x.cfg file into your SNES9x folder (if there already is one, overwrite it), and make sure there are 2 sub-folders named "Roms" and "Saves". Place your patched ROM in the Roms folder and place the provided savestate, MarioPaint.000, in the Saves folder. **Make sure both your patched ROM and the savestate have the same filename. If they don't, rename one or both of them so that they match.**

7. Open SNES9x and load the ROM (Ctrl + O, or File > Open ROM...), then load the supplied savestate (Press the F1 key, or go to File > Load Game > Slot #0). You should now be on a blank canvas with the stamps page open at the top, with each stamp set to a single colored pixel.

8. Load your chosen Lua script into SNES9x (Press spacebar twice, or go to File > Lua Scripting > New Lua Scripting Window..., then choose your file from the output folder), then press Run and watch the image draw! *(Note: you can press the M key to toggle the sound.)* Do not close the scripting window or the drawing will stop. The drawing is complete once the game returns to normal speed. 


Voila! Your automatic work of art is finished!


From here, you can take screenshots (press the F12 key; screenshots are saved in the SNES9x Screenshots folder by default) or create a new savestate to keep your artwork (File > Save Game), but I don't recommend saving over Slot #0. You can also perform a hard save by clicking the robot head in the bottom center and pressing save. Remember, this modified ROM uses a controller instead of the mouse for inputs. By default, the arrow keys move your cursor and the D key selects, and you can hold the A or S keys to move your cursor faster. Once the robot is done saving your file (you can speed this up by holding Tab), it will create a .srm file in your Saves folder. This .srm is compatible with unmodified Mario Paint ROMs, and you can even load it onto a flash cart, or onto an original Mario Paint cartridge by using a Retrode, a RetroBlaster, or similar such devices. 

**We hope you enjoy making lots of Mario Paint masterpieces!**

## Additional image settings

Included is a file called config.txt. This gives you extra control over how autoMP processes your input images. Change the number next to the setting name to change that setting.
* palette_select - Allows you to limit the colors your image uses.
  *  0 - Default. Uses every color in the Mario Paint palette.
  *  1 - Greyscale. Turns your image black and white.
  *  2 and onward - Your custom palettes - more info below!
* scaling - Alters the image size to better fit it into the canvas space.
  * 0 - No effect. Retains the shape of your image.
  * 1 - Zoom. Crops the edges of your image until it fills the canvas.
  * 2 - Stretch. Makes your image taller or wider until it fills the canvas.
* dither - Turn image dithering on or off.
  * 0 - Dithering off.
  * 1 - Dithering on.

### Custom palettes
To create your own custom palettes, you can fill out the blank templates in config.txt, then make your own following the same template. You can name your palettes whatever you like, but I recommend numbering them for easier selection. Below the palette name, type "colors = " followed by the colors you want to use, **seperated by commas but no spaces.**\
The available colors are as follows:\
<img src="mppalette.png" width="400">\
Type the color names exactly as listed here or they'll be skipped.\
**You can create as many custom palettes as you want, so please have fun experimenting with them!**

*Note: each time you run autoMP, it will apply the current settings to every image in the input folder, even images that have been processed already. This will overwrite the existing preview images and scripts in the output folder. You may want to store your images and scripts in a seperate folder to prevent them from being overwritten.*
## Q&A

*Q: Why do I have to load that savestate?*

A: The stamps! Mario Paint's smallest brush size is 2x2. In order to work around this and paint with a single pixel brush, we need a custom stamp for each color in the palette which is, you guessed it, only a single pixel. To save you (and the bot) from having to make the stamps every time you want to print, you can simply load the savestate with all the stamps set up and the cursor in position, ready to draw. Plus, it saves you the trouble of catching Mario on the title screen!

*Q: What is the included snes9x.cfg file for?*

A: The .cfg or configuration file saves the emulator's settings. You can skip this step if you want, but I've disabled some obtrusive on-screen text, mapped the spacebar to open the Lua Script Window and the M key to toggle sound, and changed the default image size to 1x, which is easily resizable if you'd like. The inclusion of this .cfg is mostly for the convenience of the layman who may not be familiar with emulators.

*Q: I don't like how my image turned out. Can I prepare an image externally and have that turned into a Lua script?*

A: Yes! You just have to make sure your image is no larger than 248x168px and that it adheres to Mario Paint's color palette. Once you have it how you like, simply put it in the input folder and run autoMP as normal. Check your preview image in the output folder, and if done correctly, it should be identical to your input image.


changelog
v1.0.0

Initial release


mariopaintFaster.lua
local imagestring = 
local imagewidth =  --max size 248
local imageheight =  --max size 168

snes9x.speedmode("turbo")

local cursorx = 0x7e0226
local cursory = 0x7e0227
local cursortype = 0x7E0426 -- 86 stamp, 84 color picker

local leftbound = 2 
local topbound = 24  
local rightbound = 250        
local bottombound = 192       

local stepsize = 1

local colorList2={'1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'}
local colorTotals={}
local colorList={}

for i=1,15 do
   colorTotals[i] = {0,colorList2[i]}
   for w in string.gfind(imagestring, colorList2[i]) do
       colorTotals[i][1] = colorTotals[i][1] + 1
   end
end

table.sort(colorTotals, function(a,b) return a[1]<b[1] end) --use < for small to large, > for large to small

colorList = colorTotals

--for centering images that do not match the aspect ratio
leftbound = math.floor(leftbound + (rightbound-imagewidth-leftbound)/2)
topbound = math.floor(topbound + (bottombound-imageheight-topbound)/2)
local colorSelected = '-'

local function paintdot(dotx,doty)
      local targetx = (dotx * stepsize) + leftbound
      local targety = (doty * stepsize) + topbound
      local currentx = memory.readbyte(cursorx)
      local currenty = memory.readbyte(cursory)
      
      while math.abs(currentx-targetx)>5 or math.abs(currenty-targety)>5 do
            if currentx < targetx then
               joypad.set(1,{right=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currentx > targetx then
               joypad.set(1,{left=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty < targety or memory.readbyte(cursortype)==84 then
               joypad.set(1,{down=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty > targety and memory.readbyte(cursortype)~=84 then
               joypad.set(1,{up=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            currentx = memory.readbyte(cursorx)
            currenty = memory.readbyte(cursory)
      end
      
      while currentx ~= targetx or currenty ~= targety do
            if currentx < targetx then
               joypad.set(1,{right=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currentx > targetx then
               joypad.set(1,{left=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty < targety or memory.readbyte(cursortype)==84 then
               joypad.set(1,{down=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty > targety and memory.readbyte(cursortype)~=84 then
               joypad.set(1,{up=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            currentx = memory.readbyte(cursorx)
            currenty = memory.readbyte(cursory)
      end
      joypad.set(1,{X=1})
      snes9x.frameadvance()
end

local function chooseColor(thecolor)
      local targetx
      local targety = 14
      if(thecolor=='1') then
           targetx = 30
      end
      if(thecolor=='2') then
           targetx = 44
      end
      if(thecolor=='3') then
           targetx = 58
      end
      if(thecolor=='4') then
           targetx = 72
      end
      if(thecolor=='5') then
           targetx = 86
      end
      if(thecolor=='6') then
           targetx = 100
      end
      if(thecolor=='7') then
           targetx = 114
      end
      if(thecolor=='8') then
           targetx = 128
      end
      if(thecolor=='9') then
           targetx = 142
      end
      if(thecolor=='A') then
           targetx = 156
      end
      if(thecolor=='B') then
           targetx = 170
      end
      if(thecolor=='C') then
           targetx = 184
      end
      if(thecolor=='D') then
           targetx = 198
      end
      if(thecolor=='E') then
           targetx = 212
      end
      if(thecolor=='F') then
           targetx = 226
      end

      local currentx = memory.readbyte(cursorx)
      local currenty = memory.readbyte(cursory)

      while math.abs(currentx-targetx)>5 or math.abs(currenty-targety)>5 do
            if currentx < targetx then
               joypad.set(1,{right=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currentx > targetx then
               joypad.set(1,{left=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty < targety  then
               joypad.set(1,{down=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty > targety  then
               joypad.set(1,{up=1,L=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            currentx = memory.readbyte(cursorx)
            currenty = memory.readbyte(cursory)
      end

      while currentx ~= targetx or currenty ~= targety or memory.readbyte(cursortype)==86 do
            if currentx < targetx then
               joypad.set(1,{right=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currentx > targetx then
               joypad.set(1,{left=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            if currenty > targety or memory.readbyte(cursortype)==86 then
               joypad.set(1,{up=1})
               snes9x.frameadvance()
               snes9x.frameadvance()
               snes9x.frameadvance()
            end
            currentx = memory.readbyte(cursorx)
            currenty = memory.readbyte(cursory)
      end
      joypad.set(1,{X=1})
      snes9x.frameadvance()
      joypad.set(1,{X=1})
      snes9x.frameadvance()
      joypad.set(1,{X=1})
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()
      snes9x.frameadvance()

      colorSelected = thecolor
end

local thisposition = 1

for k = 1,15 do
   snes9x.message(colorTotals[k][2])
   chooseColor(colorTotals[k][2])

   for i = 0,imageheight-1 do
       for j = 0,imagewidth-1 do
               thisposition = i*imagewidth + j + 1

               curPix = string.sub(imagestring,thisposition,thisposition)

               if curPix == colorTotals[k][2] then
                  paintdot(j,i)
               end
       end
   end
end


autoMP.py
from PIL import Image
import os
import os.path
from pathlib import Path
import configparser
from PIL import ImageOps

config=configparser.ConfigParser(allow_no_value=True)
if os.path.isfile('.\config.txt')==False:
#default settings stored here. will generate a new config file with default settings if it does not exist
    config.add_section('settings')
    config.set('settings','# to restore all settings to default, move or delete config.txt and run autoMP.py')
    config.set('settings', 'palette_select', '0')
    config.set('settings','# 0 - default, 1 - greyscale, 2 and up - custom palettes')  
    config.set('settings', 'scaling', '0')
    config.set('settings','# 0 - original, 1 - zoom, 2 - stretch')  
    config.set('settings', 'dither', '1')
    config.set('settings','# 0 - off, 1 - on')  
    #config.set('settings','')  #config template pair, first is description comment
    #config.set('settings', '', '')
    config.add_section('1. greyscale')
    config.set('1. greyscale','colors','black,grey,silver,white')
    config.set('1. greyscale','---------------#custom palettes#---------------') 
    config.set('1. greyscale','# available colors:')
    config.set('1. greyscale','# red, orange, yellow, lime, green, cyan, blue, rust, brown, tan, magenta, black, grey, silver, white')
    config.set('1. greyscale','# separate colors with commas but no spaces')
    config.add_section('2. ')
    config.set('2. ','colors =')
    config.add_section('3. ')
    config.set('3. ','colors =')
    config.add_section('4. ')
    config.set('4. ','colors =')
    with open(r".\config.txt", 'w') as configfile:
        config.write(configfile)
#future config file options:
#palette white skip
#fill canvas with colored border - pad(img, (248,168))

#import settings from config.txt
config.read(".\config.txt")
settings=config['settings']
palette_select=int(settings['palette_select'])
scaling_select=int(settings['scaling'])
config_sections=config.sections()
dither_select=int(settings['dither'])
#available colors
colors={
'red':[255,0,0],
'orange':[255,130,0],
'yellow':[255,251,0],
'lime':[0,251,0],
'green':[0,130,66],
'cyan':[0,251,255],
'blue':[0,0,255],
'rust':[198,65,33],
'brown':[132,97,0],
'tan':[255,195,132],
'magenta':[198,0,198],
'black':[0,0,0],
'grey':[132,130,132],
'silver':[198,195,198],
'white':[255,251,255]}

#import custom palette settings
if palette_select>0:
    testp=config[config_sections[palette_select]]
    testp=testp['colors']
    testp=testp.split(',')
    custom={x: y for (x,y) in colors.items() if x in testp}

mp_palette=colors
#palette prep
if palette_select>0:
    mp_palette=custom
characters=mp_palette
mp_palette=[x for l in mp_palette.values() for x in l]
numcolors=int(len(mp_palette)/3)
padding=mp_palette[0:3]*(256-numcolors)
mp_palette=mp_palette+padding

#the palette needs to be padded with duplicate colors up to 256 to work correctly :)
#we use the first color to ensure no extra colors are added
#this whole section could be cleaned up a lot :/

# a palette image to use for quantization
pimage = Image.new("P", (1, 1), 0)
pimage.putpalette(mp_palette)

#prep specific palette for drawing routine
mpcolors=['1','2','3','4','5','6','7','8','9','A','B','C','D','E','F']
check=list(characters)
mpfinal=dict(zip(list(colors),mpcolors))
mpcolors=[mpfinal[x] for x in check]

#locating all relevant image files in the input folder
dir_path = os.path.dirname(os.path.realpath(__file__))
fin=dir_path+"\input"
fout=dir_path+"\output\\"
for file in Path(fin).iterdir():

    #converting image to correct resolution
    image = Image.open(file) 
    image=image.convert("RGB")
    if scaling_select==0:
        image.thumbnail((248,168))
    elif scaling_select==1:
        image=ImageOps.fit(image, (248, 168))
    elif scaling_select==2:
        image=image.resize((248,168))
    else:
        image.thumbnail((248,168))

    #applying correct palette
    image=image.quantize(colors=numcolors, palette=pimage, dither=dither_select)

    #saving image preview
    head, tail = os.path.split(file)
    root, ext=os.path.splitext(tail)
    image.save(fout+root+".bmp")

    pix = image.load()
    outstring = ""

    #generating output string
    for i in range(0,image.size[1]):
        for j in range(0,image.size[0]):
        
           outstring+= mpcolors[pix[j, i]]
           
    #generating a new lua file
    file=open('mariopaintFaster.lua','r')
    lines=file.readlines()
    lines[0]="local imagestring = \"%s\"\n" %(outstring)
    lines[1]="local imagewidth = %d\n" %(image.size[0])
    lines[2]="local imageheight = %d\n" %(image.size[1])

    file=open(fout+root+".lua",'w')
    file.writelines(lines)
    file.close()
